js/libs/form.js

```diff
@@ -472,185 +474,242 @@ Fliplet.Widget.instance('form-builder', function(data) {
+         *  if its empty string show default message
+         * @returns {void} shows a toast message to users
+         */
+        function showValidationMessage(errorMessage) {
+          errorMessage = errorMessage || 'Please complete all required fields.';
+          Fliplet.UI.Toast(errorMessage);
+        }
 
-        function appendField(name, value) {
-          if (Array.isArray(formData[name])) {
-            formData[name].push(value);
-          } else if (typeof formData[name] !== 'undefined') {
-            formData[name] = [formData[name], value];
-          } else {
-            formData[name] = value;
-          }
+        /**
+         * This method will decide what we will do after isFormInvalid hook
+         * 
+         * @returns {Promise} With this logic:
+         *  1. In case there was no listener on isFormInvalid hook we will show the toast
+         *     with default message
+         *  2. In case when listener was resolved with Promise.resolve() we allow user to submit invalid form
+         *     and will not show the toast
+         *  3. In case when listener was resolved with Promise.reject() we will not show the toast
+         *     and not allow form to submit
+         *  4. In case when listener was resolved with Promise.reject('') we will show the toast with default text
+         *  5. In case when listener was resolved with Promise.reject('error text') we will show the toast
+         *     with 'error text' message
+         */
+        function onFormInvalid() {
+          return new Promise(function(resolve, reject) {
+            Fliplet.Hooks.run('isFormInvalid', invalidFields)
+              .then(function(result) {
+                if (!result.length) {
+                  showValidationMessage('');
+
+                  return reject();
+                }
+                
+                return resolve();
+              })
+              .catch(function(response) {
+                switch (typeof response) {
+                  case 'string':
+                    return showValidationMessage(response);
+                  default:
+                    return reject();
+                }
+              })
+          })
         }
 
-        var errorFields = Object.keys(this.errors);
-        var fieldErrors = [];
-        if (errorFields.length) {
-          errorFields.forEach(function (fieldName) {
-            fieldErrors.push(errorFields[fieldName]);
-          });
+        function onFormSubmission() {
+          if (!$vm.isFormValid) {
+            return onFormInvalid();
+          }
 
-          $vm.error = fieldErrors.join('. ');
-          $vm.isSending = false;
-          return;
+          return Promise.resolve();
         }
 
-        this.fields.forEach(function(field) {
-          var value = field.value;
-          var type = field._type;
+        return onFormSubmission()
+        .then(function() {
```

Comment:
This indentation is wrong. Either:

```suggestion
        return onFormSubmission()
          .then(function() {
```

...or...

```suggestion
        return onFormSubmission().then(function() {
```

Then make sure the rest of this promise chain follows the same pattern.

Reply:
@tonytlwu Fixed